internal/ast/node.go

```diff
@@ -0,0 +1,189 @@
+package ast
+
+import (
+	"github.com/cedar-policy/cedar-go/types"
+)
+
+type Node struct {
+	v IsNode // NOTE: not an embed because a `Node` is not a `node`
+}
+
+func NewNode(v IsNode) Node {
+	return Node{v: v}
+}
+
+func (n Node) AsIsNode() IsNode {
+	return n.v
+}
+
+type StrOpNode struct {
+	Arg   IsNode
+	Value types.String
+}
+
+func (n StrOpNode) isNode() {}
+
+type BinaryNode struct {
+	Left, Right IsNode
+}
+
+func (n BinaryNode) isNode() {}
+
+type NodeTypeIfThenElse struct {
+	If, Then, Else IsNode
+}
+
+func (n NodeTypeIfThenElse) isNode() {}
+
+type NodeTypeOr struct{ BinaryNode }
+
+type NodeTypeAnd struct {
+	BinaryNode
+}
+
+type NodeTypeLessThan struct {
+	BinaryNode
+}
+type NodeTypeLessThanOrEqual struct {
+	BinaryNode
+}
+type NodeTypeGreaterThan struct {
+	BinaryNode
+}
+type NodeTypeGreaterThanOrEqual struct {
+	BinaryNode
+}
+type NodeTypeNotEquals struct {
+	BinaryNode
+}
+type NodeTypeEquals struct {
+	BinaryNode
+}
+type NodeTypeIn struct {
+	BinaryNode
+}
```

Comment:
Can I assume that these types exist, rather than a `BinaryOpNode`, for optimization purposes, e.g. [Monomorphization](https://en.wikipedia.org/wiki/Monomorphization)?

Reply:
I'd say
- optimization maybe
- code ergonomics, it makes things stronger typed, so instead of two possible enum problems (the NodeType and the NodeBinaryInternalType) we just have the NodeType to deal with.